dagshub/common/api/repo.py

```diff
@@ -259,78 +260,134 @@ def get_storage_file(self, path: str) -> bytes:
             raise RuntimeError(error_msg)
         return res.content
 
-    def _get_files_in_path(self, path, revision=None, recursive=False) -> List[ContentAPIEntry]:
+    def _get_files_in_path(
+        self, path, revision=None, recursive=False, traverse_storages=False
+    ) -> List[ContentAPIEntry]:
         """
         Walks through the path of the repo, returning non-dir entries
         """
 
-        dir_queue = [path]
+        dir_queue = []
         files = []
 
+        list_fn_folder = partial(self.list_path, revision=revision)
+        list_fn_storage = self.list_storage_path
+
+        def push_folder(folder_path):
+            dir_queue.append((folder_path, list_fn_folder))
+
+        def push_storage(storage_path):
+            dir_queue.append((storage_path, list_fn_storage))
+
+        # Initialize the queue
+        is_storage_path = str(path).lstrip("/").split("/")[0] in {"s3:", "gs:", "azure:"}
+        if is_storage_path:
+            # Handle storage paths - they start with s3:/, gs:/ or azure:/
+            path = str(path).replace(":", "", 1)
+            push_storage(path)
+        else:
+            push_folder(path)
+
         progress = get_rich_progress(rich.progress.MofNCompleteColumn())
         task = progress.add_task("Traversing directories...", total=None)
 
-        def step(step_path):
-            res = self.list_path(step_path, revision)
+        def step(step_path, list_fn):
+            """
+            step_path: path in the repo to list
+            list_fn: which function to use to list it (can be list_path or list_storage_path)
+            """
+            res = list_fn(step_path)
             for entry in res:
                 if entry.type == "file":
                     files.append(entry)
-                elif entry.type == "dir" and recursive:
-                    dir_queue.append(entry.path)
-                # TODO: storage
+                elif recursive:
+                    if entry.versioning == "bucket":
+                        if traverse_storages:
+                            push_storage(entry.path)
+                    else:
+                        push_folder(entry.path)
             progress.update(task, advance=1)
 
         with progress:
             while len(dir_queue):
-                query_path = dir_queue.pop(0)
-                step(query_path)
+                query_path, list_fn = dir_queue.pop(0)
+                step(query_path, list_fn)
 
         return files
 
     def download(
-        self, remote_path, local_path=".", revision=None, recursive=True, keep_source_prefix=True, redownload=False
+        self,
+        remote_path,
+        local_path=".",
+        revision=None,
+        recursive=True,
+        keep_source_prefix=False,
+        redownload=False,
+        download_storages=False,
```

Comment:
I think you forgot to expose this in the CLI

Reply:
Right, thank you, totally forgot